rosette/rbl/rosette/namespace-util.rbl

```diff
@@ -0,0 +1,119 @@
+;;; TODO: Remove unused methods in Tuplespace and add tests for the rest
+
+;;; TODO: Add tag for whether persistent
+;;; TODO: Translate match in compiler
+
+(random-number-init)
+;;; The debugSymbol is inserted so that the output is easier to read
+(defOprn generateFresh)
+(defProc (generateFresh debugSymbol) (->symbol (string-concat "Rholang-" debugSymbol "-" (->string (random-number)))))
+
+(defOprn append)
+(defPure Tuple (append item) (reverse [item & (reverse (self))]))
+
+(defOprn string-length<)
+(defOprn string-length>=)
+(defPure String (string-length< other) (< (string-length (self)) (string-length other)))
+(defPure String (string-length>= other) (not (string-length< (self) other)))
+
+(defOprn symbol-length<)
+(defOprn symbol-length>=)
+(defPure Symbol (symbol-length< other) (string-length< (->string (self)) (->string other)))
+(defPure Symbol (symbol-length>= other) (string-length>= (->string (self)) (->string other)))
+
+;;; Deletes element at position i in Tuple
+(defOprn delete-i)
+(defPure Tuple (delete-i p)
+    (concat (prim-sub-object (self) 0 p)
+        (prim-sub-object (self) (fx+ p 1)
+                 (fx- (prim-size (self)) (fx+ p 1)))))
+
+(defOprn replace-i)
+(defPure Tuple (replace-i p item)
+    (concat
+        (prim-sub-object (self) 0 p)
+        [item]
+        (prim-sub-object (self) (fx+ p 1)
+                 (fx- (prim-size (self)) (fx+ p 1)))))
+
+(defOprn symbol-list-concat)
+(defPure Tuple (symbol-list-concat) (->symbol (fold (self) (proc [e acc k] (k (string-concat acc (->string e)))) "")))
+
+(defOprn flatten)
+(defPure Tuple (flatten) (fold (self) (proc [e acc k] (k (tuple-concat e acc))) []))
```

Comment:
Needs a pretty specific argument, a tuple containing tuples, otherwise it gets upset.
```
rosette> (flatten [['a ['b]] 'c 'd ['e 'f 'g] 'h])
*** 1st argument is not a Tuple: ((FO Prim tuple-concat) c ['a ['b]])
 in (proc [e acc k] (k (tuple-concat e acc)))
```
Proper (?) arguments flatten in little-endian tuple-order:
```
rosette> (flatten [['a 'b] ['c 'd] ['e 'f 'g]])
['e 'f 'g 'c 'd 'a 'b]
```
I would either rename it to `append-tuples` or something, or use the more standard (untested)
```
(defProc (flatten lst)
  (cond ((null? lst) [])
        ((not (type? lst Tuple))
         (new Tuple lst))
        (else (fold lst
                    (proc [e acc k]
                      (k (tuple-concat acc (flatten e))))
                    []))))
```
for
```
rosette> (flatten [['a ['b]] 'c 'd ['e 'f 'g] 'h])
['a 'b 'c 'd 'e 'f 'g 'h]
```

Reply:
Will use

```
(defProc (flatten lst)
  (cond ((null? lst) [])
        ((not (type? lst Tuple)) [lst])
        (else
          (fold lst
            (proc [e acc k]
              (k (tuple-concat acc (flatten e))))
            []))))
```

Reply:
Hmm the following causes a test to fail.
```
(defOprn flatten)
(defProc (flatten lst)
  (cond ((null? lst) [])
        ((not (type? lst Tuple)) [lst])
        (else
          (fold lst
            (proc [e acc k]
              (k (tuple-concat acc (flatten e))))
            []))))
```
I want to go to sleep now so I think I'll just keep the old flatten to now and add a TODO: comment.

Reply:
Oh nvm I forgot I was actually going for a depth of 1 flatten (if that makes sense). I'll go for the rename.

Reply:
And flip the order.